src/combinatorics/simple_graph/matching.lean

```diff
@@ -0,0 +1,74 @@
+/-
+Copyright (c) 2020 Alena Gusakov. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Alena Gusakov.
+-/
+import data.fintype.basic
+import data.sym2
+import combinatorics.simple_graph.basic
+/-!
+# Matchings
+
+
+## Main definitions
+
+* a `matching` on a simple graph is a subset of its edge set such that
+   no two edges share an endpoint.
+
+* a `perfect_matching` on a simple graph is a matching in which every
+   vertex belongs to an edge.
+
+TODO:
+  - Lemma stating that the existence of a perfect matching on `G` implies that
+    the cardinality of `V` is even (assuming it's finite)
+  - Hall's Marriage Theorem
+  - Tutte's Theorem
+  - consider coercions instead of type definition for `matching`:
+    https://github.com/leanprover-community/mathlib/pull/5156#discussion_r532935457
+  - consider expressing `matching_verts` as union:
+    https://github.com/leanprover-community/mathlib/pull/5156#discussion_r532906131
+
+TODO: Tutte and Hall require a definition of subgraphs.
+-/
+open finset
+universe u
+
+namespace simple_graph
+variables {V : Type u} (G : simple_graph V)
+
+/--
+A matching on `G` is a subset of its edges such that no two edges share a vertex.
+-/
+structure is_matching (M : set (sym2 V)) : Prop :=
+(sub_edges : M ⊆ G.edge_set)
+(disjoint : ∀ (x y ∈ M) (v : V), v ∈ x ∧ v ∈ y → x = y)
+
+/--
+`matching G` is the type of matchings over `G`.
+-/
+def matching : Type u := {M : set (sym2 V) // G.is_matching M}
```

Comment:
Something you might want to consider later is whether you need both the prop and the subtype
```suggestion
/--
A matching on `G` is a subset of its edges such that no two edges share a vertex.
-/
structure matching :=
(carrier : set (sym2 V))
(sub_edges : carrier ⊆ G.edge_set)
(disjoint : ∀ (x y ∈ carrier ) (v : V), v ∈ x ∧ v ∈ y → x = y)
```

The advantage of this spelling is you can use `matching.sub_edges` instead of `matching.prop.sub_edges`

Reply:
I do like this, I'm adding it to my next PR. Thanks!

Reply:
I'd prefer calling it `edges` instead of `carrier`, but this seems like a good idea.  (You can also implement `has_mem` so that the matching "is" a set of edges if that makes things more convenient.)